### ISSUE TITLE
Add private metadata to documents

### ISSUE THREAD

**Issue by knowtheory** [documentcloud/documentcloud#333]
This proposal is to add a second field to Documents where private data can be stored (and independently searched) to documents for use by teams organizing documents with workflow metadata.

**Comment by reefdog** [comment]
Is this "private to team" (aka exposed to privileged users, like tagging with private attributes) or "private/internal for DC use" (for the sort of freeform metadata that doesn't warrant a column on the table, like ingestion method)?

**Comment by knowtheory** [comment]
I'd intended as the former, with attributes left by DocumentCloud the system, but you're probably right to point out that the integrity of data left by DocumentCloud might warrant a different data structure.  (not super excited about THREE metadata fields tho :P )

**Comment by reefdog** [comment]
Right. There's an easy argument for the cost:benefit of the public and private metadata fields. But that internal one could be a slippery slope. On the one hand, I like having a place to store stuff (like ingestion method) that will be null on 99% of objects. On the other hand, even with JSONB, I have to imagine this metadata should be treated as difficult to parse/search. I trust your instincts.

**Comment by jashkenas** [comment]
Just to throw in two cents — it seems like DocumentCloud-specific metadata would always be better handled by real DB fields. It's your DB, so why not use it? HStore/JSON columns are always going to be more difficult to search/update.

As for a second Document "arbitrary metadata" field, it sounds like a really advanced feature that might only be used by a tiny fraction of newsrooms. Is there a real-world use case you're thinking of in particular? For a team advanced enough to use this feature — can this really not be better handled by that team maintaining a metadata DB of their documents on their own end? In conjunction with the API?

**Comment by knowtheory** [comment]
Yep!  There are two use cases.

The less compelling of the two are folks who are trying to use DocumentCloud as their canonical private data store for documents via API access and who don't want to have to track syncing issues across two stores (especially if they provide 3rd parties access to their data).

The more compelling case are for teams who need to track metadata on documents so that they can collaborate privately (e.g. "this document was is assigned to, or was reviewed by X person") but where the information should not be made public.  This is operationally important where folks want to be able to tag documents and provide search links to other collaborators without hand curating a number of project folders (i've discussed this with a few teams in the past including the AP)

**Comment by jashkenas** [comment]
> The more compelling case are for teams who need to track metadata on documents so that they can collaborate privately (e.g. "this document was is assigned to, or was reviewed by X person") but where the information should not be made public. This is operationally important where folks want to be able to tag documents and provide search links to other collaborators without hand curating a number of project folders (i've discussed this with a few teams in the past including the AP)

Neat! Sounds like a perfect use case. Good luck making the distinction between the two metadata stores clear in the UI ;)

(Maybe just "Edit Private Document Data" in the drop-down, using the same edit dialog, is a good place to start)
